Draft: implement super-ellipsoid particle shapes for use in granular simulations

jibril-b-coulibaly opened the pull request:
**Summary**

This PR allows super-ellipsoid particle shapes and, contact detection and force calculation for granular simulations.

**Related Issue(s)**

follow-up #3999 

**Author(s)**

Jibril B. Coulibaly (Sandia National Laboratories)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

- Super-ellipsoids require 2 additional exponents that will be read from and written to the data file. For `read_data`, I believe they can be made optional and exponents values default to 2. `write_data` would always write these 2 additional values, which shouldn't be a problem for LAMMPS. This could change the behavior of some post-processing routine that is not expecting the 2 additional values on the ellipsoid lines
- more TBD

**Implementation Notes**

- Super-ellipsoid particles are defined by adding blockiness exponents to the `ellipsoid` atom style.
- The algorithms and methods are mostly based on the paper by [Podlozhnyuk et al. 2017](https://link.springer.com/article/10.1007/s40571-016-0131-6) and can/will leverage some of the existing LIGGGHTS implementation.



**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

TBD

Commit:
create data structure for super-ellipsoid inside bonus

Commit:
add functions to compute the circumscribed radius of the super ellipsoid.
add function to set blockiness and modify set_shape function to include circumscribed radius.

Also add function to compute volume of the super ellipsoid but this does not compile because
std::beta is not recognized. We need a way to compute the beta function, either from existing library
or by implementing it ourselves, e.g., in math_special, TODO

jibril-b-coulibaly commented:
Hello, I'm running into an obstacle I'd like your insight about.

The volume and moments of inertia of the super-ellipsoid are defined in terms of the Beta function.

The [STL special functions](https://en.cppreference.com/w/cpp/numeric/special_functions) are implementation-dependent and `std::beta()` does not compile on my end: I need to include `<tr1/cmath>` and call `std::tr1::beta()` as suggested by the doc. TR1 is from C++03 so it seems like it should work on most implementation, but it does not pass CI. For comparison, the [LIGGGHTS implementation](https://github.com/CFDEMproject/LIGGGHTS-PUBLIC/blob/bbd23c8521fe0a4f3f5bfd0d3dc9a7dddd68d90b/src/math_extra_liggghts_superquadric.cpp#L48) checks:

```
#if __STDCPP_MATH_SPEC_FUNCS__ >= 201003L
    #define BETA_NAMESPACE std
#elif defined(HAVE_TR1_CMATH)
    #include <tr1/cmath>
    #define BETA_NAMESPACE std::tr1
#elif defined(BOOST_INCLUDED)
    #include "boost/math/special_functions/beta.hpp"
    #define BETA_NAMESPACE boost::math
```

Would a similar check for `__STDCPP_MATH_SPEC_FUNCS__ >= 201003L` be appropriate in`atom_vec_ellipsoid.cpp` ? Or would special functions in LAMMPS need to be managed at a higher level, e.g., `math_special`?

Thank you

Commit:
add blockiness to pack/unpack methods

modify std::beta to tr1::std::beta (forgotten in previous commit)

akohlmey commented:
> Would a similar check for `__STDCPP_MATH_SPEC_FUNCS__ >= 201003L` be appropriate in`atom_vec_ellipsoid.cpp` ? Or would special functions in LAMMPS need to be managed at a higher level, e.g., `math_special`?

As far as I can tell `std::beta()` is part of the C++17 standard. But core code has to be compatible with C++11. For that we have to provide our own implementation. We prefer to keep the code itself as clean as possible, so a wrapper function is the way to go. The MathSpecial namespace would be the best match. Please also note, that we (very deliberately) do not use Boost.

jibril-b-coulibaly commented:
> For that we have to provide our own implementation. We prefer to keep the code itself as clean as possible, so a wrapper function is the way to go.

For the wrapper function, do I understand correctly that I would need to write a copy of an existing implementation, e.g., the way `erfcx_y100` is [implemented in math_special](https://github.com/lammps/lammps/blob/30051ec260512df8da9315c49a79da962018cf99/src/math_special.cpp#L200) ?

stanmoore1 commented:
@jibril-b-coulibaly seeing some compile errors with this PR:
```
/mnt/lammps/workspace/dev/pull_requests/compilation_tests/lammps/src/atom_vec_ellipsoid.cpp: In member function 'virtual void LAMMPS_NS::AtomVecEllipsoid::pack_data_pre(int)':
/mnt/lammps/workspace/dev/pull_requests/compilation_tests/lammps/src/atom_vec_ellipsoid.cpp:513:5: error: 'block' was not declared in this scope
  513 |     block = bonus[ellipsoid_flag].block;
      |     ^~~~~
/mnt/lammps/workspace/dev/pull_requests/compilation_tests/lammps/src/atom_vec_ellipsoid.cpp:514:5: error: 'flag_super' was not declared in this scope
  514 |     flag_super = bonus[ellipsoid_flag].flag_super;
      |     ^~~~~~~~~~
/mnt/lammps/workspace/dev/pull_requests/compilation_tests/lammps/src/atom_vec_ellipsoid.cpp: In member function 'double LAMMPS_NS::AtomVecEllipsoid::compute_radcirc(double*, double*, bool)':
```

Commit:
add omitted declarations

Commit:
add blockiness parameter to read_data and write_data.

reading blockiness parameters is made optional for backward compatibility.

Commit:
implement Beta function in MathSpecial based on cephes/scipy implementation.

This is necessary for C++11 compatibility since std::beta() is C++17

Implementation is modified/simplified assuming strictly positive arguments (always the case for super-ellipsoid calculations involving Beta)

Commit:
fix logic for super-ellipsoid flag

Commit:
block and flag_super must be initialized with set_shape when ellipsoid[i]<0

Commit:
allow set command to set the ellipsoid shape.

Commit:
allow shape and block to be set in any order by setting shape to 1 1 1 if block is called before shape.

Explain that default behavior in the doc. add example and line about atom-style variable for shape and block

Commit:
add unit test for new bonus properties to test_atom_styles

Commit:
implement moment of inertia for super ellipsoid. Make block and flag arguments optional to keep existing calls as is

Commit:
Create principal inertia function: avoids code duplication and fixes mistakes.

Did not change INTEL and GPU functions yet because of different, special syntax.

Commit:
Add principal moments of inertia to bonus struct to avoid recomputing at every step.

Add moments of inertia calculation to set_shape, set_block and set_density for ellipsoids

Move volume calculation to MathExtra

Commit:
correct bug in inertia, forgotten 0.5 factor

Commit:
fix omitted pointer initialization

Commit:
inertia in set DENSITY must be computed for super-ellipsoid, not default arguments

Commit:
use radius pointer to store circumscribed radius of ellipsoid. Move radius calculation to MathExtra

akohlmey commented:
@jibril-b-coulibaly can you please give us an update on this pull request / project? I has been sitting idle for a long time now and accumulated a bunch of merge conflicts since.
If this is not going to be completed soon, would you mind closing it and then resubmitting when it is closer to completion?

jibril-b-coulibaly commented:
> @jibril-b-coulibaly can you please give us an update on this pull request / project? I has been sitting idle for a long time now and accumulated a bunch of merge conflicts since. If this is not going to be completed soon, would you mind closing it and then resubmitting when it is closer to completion?

Hi @akohlmey ,
I had planned to complete this work by the end of 2024 but had far less time than anticipated. I will close it and re-submit it after I fix the conflicts and complete it on my own branch
